### DISCUSSION TITLE
More batteries included

### DISCUSSION THREAD

**Discussion by andrewbaxter** [VirtusLab/scala-cli#3014] [Ideas]
Bouncing here from Ammonite, and I thought it might bear mention.

My use case for scripting is stuff like operating system glue and local automation, like boot scripts setting up mount points or device links, or analyzing packet captures when I switch ISPs and my internet inexplicably stops working on my old router.

A huge part of why I'm using Python instead of anything else is that it's batteries included with command line parsing, json/base64, http requests (as painful as they are), subprocess management, etc.  Is there any way this could be supported here?  Like a "scala-cli-plus" with a bunch of common libraries bundled?

FWIW _which_ libraries aren't really an issue, I'm using some pretty terrible standard library Python things because what matters there is not using bash... for the same reason I'd use Scala CLI instead of Python.  Pre-creating a cache isn't really an option, or at least if I'm going to do something like that I might as well just compile.

**Comment by bishabosha** [comment]
did you know about the `--toolkit` option/directive? it adds [scala toolkit](https://docs.scala-lang.org/toolkit/introduction.html) to the classpath - including libraries for processes/io/http-requests

**Reply by andrewbaxter** [reply]
Thanks! I didn't know about that, that was the missing piece in my search!  AFAICT "scala toolkit" doesn't include any commandline parser tools, but the "typelevel toolkit" does (I think?).  The typelevel toolkit looks really cool.

Liveblogging a here a bit, but:

The documentation you linked doesn't explain how to get it (aside from as a remote dependency)... since it's "batteries included" I thought it might be included either in Scala itself or in Scala-CLI but I couldn't find any documentation confirming that.  I installed both on my system but didn't see it in any of the installed jars. Neither Arch linux nor NixOS have "scala toolkit" in their list of packages (nor typelevel toolkit, but I thought I'd start by searching for scala toolkit).

The `scala-cli` documentation just says this about the option:
> Add toolkit to classPath (not supported in Scala 2.12), 'default' version for Scala toolkit: 0.4.0, 'default' version for typelevel toolkit: 0.1.23
I'm not sure how to read that.  Is the "default" what you get if you don't specify `--with-toolkit`, so does that mean it's packaged with `scala-cli`? Or does it mean you need to give it the path to those toolkits which match the versions listed there?

I was going to try it with
```
#!/usr/bin/env -S scala-cli shebang --offline --scala-version 2.13
//> using toolkit typelevel:default
```
but that fails immediately with
```
Unrecognized argument: The `--offline` option is experimental.
You can run it with the `--power` flag or turn power mode on globally by running:
  scala-cli config power true.

To list all available options, run
  scala-cli shebang --help
```

Modifying it to
```
#!/usr/bin/env -S scala-cli shebang --power --offline --scala-version 2.13
//> using toolkit typelevel:default
```
I now get
```
The `--offline` option is experimental
Please bear in mind that non-ideal user experience should be expected.
If you encounter any bugs or have feedback to share, make sure to reach out to the maintenance team at https://github.com/VirtusLab/scala-cli
[error]  Error downloading org.scala-lang:scala-compiler:2.13
  not found: /home/.../.ivy2/local/org.scala-lang/scala-compiler/2.13/ivys/ivy.xml
  not found: /home/.../.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-compiler/2.13/scala-compiler-2.13.pom
  not found: /home/.../.cache/scalacli/local-repo/1.2.1/org.scala-lang/scala-compiler/2.13/ivys/ivy.xml
  No fallback URL found
```

I looked for a way to point it to a local scala installation but the only flags I found that seemed maybe relevant just said `[Internal]` in the docs and my scala installation doesn't have any xml files so that seems like a dead end.

I feel like I'm working at odds with `scala-cli` at this point, although this is so close to what I need.

**Reply by andrewbaxter** [reply]
I just tried it without --offline to see what it did, and it started downloading the toolkit.  So I guess the default versions listed in `--with-toolkit` are what it downloads if you _don't_ specify `--with-toolkit`, and `--with-toolkit` is a way to avoid the download?

Unfortunately I don't think this makes the distributing-offline-scripts story particularly easier...  in order to use a script on a system I need to make sure `scala-cli` is available plus `scala`, plus the toolkit jar, and make sure all the scripts include shebang flags to pass in the path of `scala` and the toolkit jar (I assume?).

**Reply by bishabosha** [reply]
So did you also read [this documentation](https://scala-cli.virtuslab.org/docs/guides/introduction/toolkit) on the Scala CLI website?

so yes scala toolkit is a maven dependency [see here](https://index.scala-lang.org/scala/toolkit/artifacts/toolkit/0.4.0). There is another discussion currently about this ["How to install scala-cli without internet"](https://github.com/VirtusLab/scala-cli/discussions/2869) - perhaps we should make a command that makes it easy to bundle all necessary libraries for a given command/project to succeed.

**Reply by bishabosha** [reply]
when scala 3.5.0 is released then its `scala` command, based on scala-cli, will bundle all the libraries to compile code with Scala 3.5.0 offline - but we didn't consider including toolkit in that offline bundle so far

**Reply by andrewbaxter** [reply]
> So did you also read [this documentation](https://scala-cli.virtuslab.org/docs/guides/introduction/toolkit) on the Scala CLI website?

No, I hadn't seen that either! I googled around for more information on the toolkit and couldn't find anything, although I guess I was probably looking more for reference type stuff.

I tried `--toolkit default` just now and same behavior AFAICT.  Is that the same as not specifying it at all?

> "How to install scala-cli without internet"

I saw that and it seemed relevant, but I guess the issue was more of a low level one in the end?

> when scala 3.5.0 is released then its `scala` command, based on scala-cli, will bundle all the libraries to compile code with Scala 3.5.0 offline - but we didn't consider including toolkit in that offline bundle so far

Oh yeah, that sounds interesting.  Would that still require a mutable system or storing the bundle somewhere?  If it'd be useful I'd be glad to make an issue.

**Reply by bishabosha** [reply]
> I tried `--toolkit default` just now and same behavior AFAICT. Is that the same as not specifying it at all?

if you don't specify that (and don't add any explicit dependency) then you shouldn't be able to use a toolkit library in code
